crates/metrics/src/label.rs

```diff
@@ -0,0 +1,545 @@
+///
+/// ```
+/// use wc_metrics::{label_name, BoolLabel};
+///
+/// type MyLabel = BoolLabel<{ label_name("my_label") }>;
+/// ```
+#[derive(Clone, Copy, Debug)]
+pub struct BoolLabel<const NAME: LabelName>(bool);
+
+impl<const NAME: LabelName> BoolLabel<NAME> {
+    /// Creates a new [`BoolLabel`].
+    pub fn new(b: bool) -> Self {
+        Self(b)
+    }
+}
+
+impl<const NAME: LabelName> From<BoolLabel<NAME>> for bool {
+    fn from(label: BoolLabel<NAME>) -> Self {
+        label.0
+    }
+}
+
+impl<const NAME: LabelName, M> DynamicLabel<M> for BoolLabel<NAME> {
+    type MetricCollection = (M, M);
+}
+
+impl<const NAME: LabelName, M> Metric for WithLabel<BoolLabel<NAME>, M>
+where
+    M: Metric,
+{
+    fn register(attrs: &Attrs) -> Self {
+        let name = const { resolve_label_name::<NAME>() };
+
+        let f = Label::from_static_parts(name, "false");
+        let t = Label::from_static_parts(name, "true");
+
+        Self {
+            collection: (
+                M::register(&attrs.with_label(f)),
+                M::register(&attrs.with_label(t)),
+            ),
+        }
+    }
+}
+
+impl<const NAME: LabelName, M> ResolveLabels<(BoolLabel<NAME>,)> for WithLabel<BoolLabel<NAME>, M>
+where
+    M: Metric,
+{
+    type Target = M;
+
+    fn resolve_labels(&self, (label,): (BoolLabel<NAME>,)) -> &M {
+        if label.0 {
+            &self.collection.1
+        } else {
+            &self.collection.0
+        }
+    }
+}
+
+/// Label with values which are unknown at the complite time.
+///
+/// Metric lookups using these labels will be slower compared to [`EnumLabel`]s
+/// and [`BoolLabel`]s. Prefer not to use these unless you really don't know
+/// your label values beforehand.
+///
+/// Despite its name the label can accept any type that implements [`ToString`],
+/// not just [`String`]s. The most frequent use-case (aside from the actual
+/// strings) - numbers.
+///
+/// Due to the lack of `&'static str` const generics at the moment the label
+/// name should be specified using the following hack:
+///
+/// ```
+/// use wc_metrics::{label_name, StringLabel};
+///
+/// type MyLabel = StringLabel<{ label_name("my_label") }>;
+/// ```
+#[derive(Clone, Copy, Debug)]
+pub struct StringLabel<const NAME: u128, T = String>(pub T);
+
+impl<const NAME: LabelName, T> StringLabel<NAME, T> {
+    /// Creates a new [`StringLabel`].
+    ///
+    /// Expects a [`Borrow`]ed version of your type as the owned version is not
+    /// needed for the label resolution.
+    pub fn new<U: ?Sized>(ref_: &U) -> StringLabel<NAME, &U>
+    where
+        T: Borrow<U>,
+    {
+        StringLabel(ref_)
+    }
+
+    /// Converts this [`StringLabel`] into the inner `T`.
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+}
+
+impl<const NAME: LabelName, T> AsRef<T> for StringLabel<NAME, T> {
+    fn as_ref(&self) -> &T {
+        &self.0
+    }
+}
+
+pub struct StringCollection<T, M: 'static> {
+    inner: ArcSwap<HashMap<T, &'static M>>,
```

Comment:
Now that you're migrating the relay, can you guesstimate the highest cardinality that such `StringCollection` can have? If it's pretty low, maybe it'd make sense to replace `HashMap` with `SmallVec`.

Reply:
Would depend on the application and whether people ab(using) this instead of using proper `EnumLabel`s.
But for the relay the only "valid" use-cases are HTTP routes and status codes, which makes the cardinality <10

Probably makes sense to implement both version, but let's try `SmallVec`

Reply:
> But for the relay the only "valid" use-cases are HTTP routes and status codes, which makes the cardinality <10

If it's `route_name * response_code` it'd likely be more than 10. Not sure if it's worth it. Can you double check the existing metrics to confirm cardinality?

Reply:
We definitely have <10 http routes in the relay

Reply:
Benchmark results: Binary search on `Vec`/`SmallVec` is 2x faster for 10 elements, and the same as `HashMap` lookup for 20 elements. 
After that the performance falls significantly, so I suggest we leave `HashMap` here, as `StringCollection` is not the recommended way to implement labels anyway and it should support all other edge-cases.